### ISSUE TITLE
Does ‘my sub f; eval 'sub f{}'’ make sense?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14303]
Migrated from [rt.perl.org#123367](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123367) (status was 'resolved')

Searchable as RT123367$

**Comment by p5pRT** [comment]
### From @cpansprout

I thought I had submitted this already\, but I couldnât find it in RT&#8203;:

I know I was the one that implemented it and made eval 'sub f\{\}' work with âmy sub fâ in scope\, because it seemed that one might expect it to work\.  But on second thought Iâm not sure it does make sense\.

âsub f \{\}â is not actually evaluated at run time\.  It is a compile\-time declaration\.  For a lexical âmyâ symbol\, it creates\, at compile time\, a protosub that later gets cloned on scope entry\.

In âmy sub f; eval 'sub f\{\}'â\, the sub declaration should create a protosub that gets cloned the \*next\* time the enclosing scope is entered\.  Right?  It least thatâs how it should work if we want it consistent\.  But this is what currently happens&#8203;:

$ \./perl \-XIlib \-e 'use feature "&#8203;:all"; for\(1\,0\) \{ my sub f; eval "sub f\{warn 42\}" if $\_; f \}'
42 at \(eval 1\) line 1\.
Undefined subroutine &f called at \-e line 1\.

Now\, Iâm not sure that this eval\-after\-my even makes that much sense\, and lexical subs are experimental\, so that is an argument for changing it\.

I \*do\* think this code makes sense\, but it doesnât work&#8203;:

$ \./perl \-Ilib \-e 'use feature "&#8203;:all"; my sub f; BEGIN \{ eval "sub f\{warn 42\}"\} f'
The lexical\_subs feature is experimental at \-e line 1\.
Undefined subroutine &f called at \-e line 1\.

BEGIN time \(when subs are being defined\) seems a logical place to decide which sub to install\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @ap

\* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-04 19&#8203;:45\]&#8203;:
> On second thought Iâm not sure it does make sense\.

You asked me to look at this a while ago\. After repeat reading\, with
some lengthy pauses included\, I still remain half\-confused about what
the problem is or what you think is the way it ought to work\.

All I can say is that to my \(possibly naÃ¯ve\) mind\, \`my sub f \{\.\.\.\}\`
ought to work just as if it said \`my $f = sub \{\.\.\.\};\`\, with subsequent
invocations of \`f\(\.\.\.\)\` being exactly equivalent to \`$f\->\(\.\.\.\)\`\. And of
course a \`sub f \{\.\.\.\}\` within scope of \`my sub f\` would likewise amount
to \`$f = sub \{\.\.\.\};\` â maybe even down to the lack of redefine warning\.

This is the only way of defining this feature that I can come up with to
produce behaviour with a coherence and self\-consistence that I will find
it graspable and predictable even in strange corner cases\.

By that line of thinking\, then\,

> $ \./perl \-XIlib \-e 'use feature "&#8203;:all"; for\(1\,0\) \{ my sub f; eval "sub f\{warn 42\}" if $\_; f \}'
> 42 at \(eval 1\) line 1\.
> Undefined subroutine &f called at \-e line 1\.

â¦ this is exactly what one would expect\. And by the same token\,

> $ \./perl \-Ilib \-e 'use feature "&#8203;:all"; my sub f; BEGIN \{ eval "sub f\{warn 42\}"\} f'
> The lexical\_subs feature is experimental at \-e line 1\.
> Undefined subroutine &f called at \-e line 1\.

â¦ this \*ought\* to work\.

Is the strict equivalence of these forms in my mind the reason I might
be thinking of this differently than you do? The following appears to
suggest it&#8203;:

> âsub f \{\}â is not actually evaluated at run time\. It is a compile\-time
> declaration\. For a lexical âmyâ symbol\, it creates\, at compile time\,
> a protosub that later gets cloned on scope entry\.

Iâm not steeped enough in the implementation to talk about it in terms
of protosub etc\, but again\, based on the strict equivalence\, \`my sub f\`
basically ought to behave essentially like \`my $f\` does WRT the compile
time/runtime effects split\.

Is this opinion useful to you?

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jan 07 22&#8203;:08&#8203;:03 2015\, aristotle wrote&#8203;:
> \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-04 19&#8203;:45\]&#8203;:
> > On second thought Iâm not sure it does make sense\.
> 
> You asked me to look at this a while ago\. After repeat reading\, with
> some lengthy pauses included\, I still remain half\-confused about what
> the problem is or what you think is the way it ought to work\.
> 
> All I can say is that to my \(possibly naÃ¯ve\) mind\, \`my sub f \{\.\.\.\}\`
> ought to work just as if it said \`my $f = sub \{\.\.\.\};\`\, with subsequent
> invocations of \`f\(\.\.\.\)\` being exactly equivalent to \`$f\->\(\.\.\.\)\`\. And
> of
> course a \`sub f \{\.\.\.\}\` within scope of \`my sub f\` would likewise
> amount
> to \`$f = sub \{\.\.\.\};\` â maybe even down to the lack of redefine
> warning\.
> 
> This is the only way of defining this feature that I can come up with
> to
> produce behaviour with a coherence and self\-consistence that I will
> find
> it graspable and predictable even in strange corner cases\.

With regular subs\, sub foo\{\} is mostly equivalent to&#8203;:

&nbsp;    BEGIN \{ \*foo = sub\{\} \}

Similarly\, state sub foo\{\} is like&#8203;:

&nbsp;    state sub foo;
&nbsp;    BEGIN \{ \\&foo = sub\{\} \}

My\-subs also have a compile\-time action\, in that the declaration creates a protosub that will be cloned at run time\, just as sub\{\.\.\.\} does\.

> By that line of thinking\, then\,
> 
> > $ \./perl \-XIlib \-e 'use feature "&#8203;:all"; for\(1\,0\) \{ my sub f; eval
> > "sub f\{warn 42\}" if $\_; f \}'
> > 42 at \(eval 1\) line 1\.
> > Undefined subroutine &f called at \-e line 1\.
> 
> â¦ this is exactly what one would expect\. And by the same token\,
> 
> > $ \./perl \-Ilib \-e 'use feature "&#8203;:all"; my sub f; BEGIN \{ eval "sub
> > f\{warn 42\}"\} f'
> > The lexical\_subs feature is experimental at \-e line 1\.
> > Undefined subroutine &f called at \-e line 1\.
> 
> â¦ this \*ought\* to work\.
> 
> Is the strict equivalence of these forms in my mind the reason I might
> be thinking of this differently than you do? The following appears to
> suggest it&#8203;:
> 
> > âsub f \{\}â is not actually evaluated at run time\. It is a compile\-
> > time
> > declaration\. For a lexical âmyâ symbol\, it creates\, at compile time\,
> > a protosub that later gets cloned on scope entry\.
> 
> Iâm not steeped enough in the implementation

Neither am I\.  I recently had a look at the code in newMYSUB\, and I donât understand it\.  This is coming from the person who \*wrote\* it\.  So that suggests that the number of people who understand that code is zero\, which is a scary thought\.

> to talk about it in terms
> of protosub etc\, but again\, based on the strict equivalence\, \`my sub
> f\`
> basically ought to behave essentially like \`my $f\` does WRT the
> compile
> time/runtime effects split\.
> 
> Is this opinion useful to you?

I donât really know the answer to that\. :\-\)

Taking this example that you say ought to work&#8203;:

> > $ \./perl \-Ilib \-e 'use feature "&#8203;:all"; my sub f; BEGIN \{ eval "sub
> > f\{warn 42\}"\} f'
> > The lexical\_subs feature is experimental at \-e line 1\.
> > Undefined subroutine &f called at \-e line 1\.

Is that equivalent to

&nbsp;    my sub f;
&nbsp;    BEGIN \{ eval '\\&f = sub \{ warn 42 \}' \}

?

If so\, then we will have aliased the sub at compile time before the lexical scope is entered\.  Since all lexical variables are implicitly reset at scope entry \(or exit\, depending on how you look at it\)\, at some point sub f has to revert back to being undefined\.

Or\, if

&nbsp;    my sub f \{ warn 41 \}
&nbsp;    BEGIN \{ sub \{ warn 42 \} \}

is equivalent to

&nbsp;    my sub f;
&nbsp;    \\&f = sub \{ warn 41 \};
&nbsp;    BEGIN \{ \\&f = sub \{ warn 42 \} \}

then the BEGIN block is a no\-op\, since the run\-time 41 assignment would take precedence\.

Similarly\, the preceding example would be equivalent to having \\&f = \<undefined sub> outside the BEGIN block\.

So the assignment model doesnât work here\.

Brad Gilbert wrote&#8203;:
> FWIW\, in Perl6 it also works&#8203;:
> 
>     $ perl6 \-e ' my &f //= sub \(\)\{???\}; f; EVAL q\[&f = sub \(\)\{say "eval"\}\]; f'
>     Stub code executed  in sub  at \-e&#8203;:1
> 
>     eval
>     $ perl6 \-e ' my &f //= sub \(\)\{???\}; f; BEGIN EVAL q\[&f = sub \(\)\{say "eval"\}\]; f '
>     eval
>     eval

I notice you are using //=\, which avoids all the problems we are dealing with here\.

What if you replace that with âmy sub f \{???\}â?

Using \\&f=\.\.\. assignment in Perl 5 I can make it work very similarly to your Perl 6 example\.  It is the declarative syntax I am asking about\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jan 07 22&#8203;:08&#8203;:03 2015\, aristotle wrote&#8203;:
> \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-04 19&#8203;:45\]&#8203;:
> > On second thought Iâm not sure it does make sense\.
> 
> You asked me to look at this a while ago\. After repeat reading\, with
> some lengthy pauses included\, I still remain half\-confused about what
> the problem is or what you think is the way it ought to work\.
> 
> All I can say is that to my \(possibly naÃ¯ve\) mind\, \`my sub f \{\.\.\.\}\`
> ought to work just as if it said \`my $f = sub \{\.\.\.\};\`\, with subsequent
> invocations of \`f\(\.\.\.\)\` being exactly equivalent to \`$f\->\(\.\.\.\)\`\. And
> of
> course a \`sub f \{\.\.\.\}\` within scope of \`my sub f\` would likewise
> amount
> to \`$f = sub \{\.\.\.\};\` â maybe even down to the lack of redefine
> warning\.
> 
> This is the only way of defining this feature that I can come up with
> to
> produce behaviour with a coherence and self\-consistence that I will
> find
> it graspable and predictable even in strange corner cases\.

Here is another way of looking at it\.  This code&#8203;:

my $f = sub \{\.\.\.\}

involves three distinct steps\.

1\. A protosub is created at compile time \(you can observe this with
&nbsp;   attributes\)\.
2\. The sub is cloned\.
3\. The clone is assigned to $f\.

With 'my sub f \{\.\.\.\}' it is clear that the scoping is identical to 'my $f = sub \{\.\.\.\}'\.

It is when the declaration precedes the definition that we can no longer us the 'my $f =' model to explain it\.

my sub f;
f\(\);
sub f \{\.\.\.\};

In this case step 1 happens at compile time\, as before\, but on line 3\.  Steps 2 and three happen on line 1\, allowing the f\(\) call on line 2 to work\.

This later definition of the sub works even if it is inside an inner sub&#8203;:

my sub f;
f\(\);
sub \{ sub f \{\.\.\.\} \}

my sub f;
f\(\);
sub BEGIN \{ sub f \{\.\.\.\} \}

It is when we have string eval that things get confusing\.

my sub f;
eval 'sub f \{\.\.\.\}';

In that code\, steps 2 and 3 happen first\, and then all three steps happen at the same time inside the eval\.  \(That the internal code that handles sub definitions should have to handle cloning as well really complicates things\.\)  This is an exception to the general rule\, which is why I question whether it was a good idea\.  Usually steps 2 and 3 happen only when the enclosing block is entered\.

When you have a BEGIN block with an eval inside it\, then things get a little screwy\.

my sub f;
BEGIN \{ eval 'sub f \{\.\.\.\}'; \}

Because of the way lexical\-sub\-definition\-in\-eval has been special\-cased to work\, you get all three steps happening at once at compile time\, and then steps 2 and 3 repeat at run time\.

Now this may all sound like internal jargon\, but it is all observable behaviour\.  Steps 2 and 3 always happen together with my\-subs\, but step 1 happens separately\.  One can observe step 1 through attributes \(i\.e\.\, based on when MODIFY\_CODE\_ATTRIBUTES is called\.\)  One can observe steps 2 and 3 \(actually a single step\) by the effects of calling the sub\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Jan 09 18&#8203;:14&#8203;:33 2015\, sprout wrote&#8203;:
> On Wed Jan 07 22&#8203;:08&#8203;:03 2015\, aristotle wrote&#8203;:
> > \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-04 19&#8203;:45\]&#8203;:
> > > On second thought Iâm not sure it does make sense\.
> >
> > You asked me to look at this a while ago\. After repeat reading\, with
> > some lengthy pauses included\, I still remain half\-confused about what
> > the problem is or what you think is the way it ought to work\.
> >
> > All I can say is that to my \(possibly naÃ¯ve\) mind\, \`my sub f \{\.\.\.\}\`
> > ought to work just as if it said \`my $f = sub \{\.\.\.\};\`\, with
> > subsequent
> > invocations of \`f\(\.\.\.\)\` being exactly equivalent to \`$f\->\(\.\.\.\)\`\. And
> > of
> > course a \`sub f \{\.\.\.\}\` within scope of \`my sub f\` would likewise
> > amount
> > to \`$f = sub \{\.\.\.\};\` â maybe even down to the lack of redefine
> > warning\.
> >
> > This is the only way of defining this feature that I can come up with
> > to
> > produce behaviour with a coherence and self\-consistence that I will
> > find
> > it graspable and predictable even in strange corner cases\.
> 
> Here is another way of looking at it\.  This code&#8203;:
> 
> my $f = sub \{\.\.\.\}
> 
> involves three distinct steps\.
> 
> 1\. A protosub is created at compile time \(you can observe this with
>    attributes\)\.
> 2\. The sub is cloned\.
> 3\. The clone is assigned to $f\.
> 
> With 'my sub f \{\.\.\.\}' it is clear that the scoping is identical to 'my
> $f = sub \{\.\.\.\}'\.
> 
> It is when the declaration precedes the definition that we can no
> longer us the 'my $f =' model to explain it\.
> 
> my sub f;
> f\(\);
> sub f \{\.\.\.\};
> 
> In this case step 1 happens at compile time\, as before\, but on line 3\.
> Steps 2 and three happen on line 1\, allowing the f\(\) call on line 2 to
> work\.
> 
> This later definition of the sub works even if it is inside an inner
> sub&#8203;:
> 
> my sub f;
> f\(\);
> sub \{ sub f \{\.\.\.\} \}
> 
> my sub f;
> f\(\);
> sub BEGIN \{ sub f \{\.\.\.\} \}
> 
> It is when we have string eval that things get confusing\.
> 
> my sub f;
> eval 'sub f \{\.\.\.\}';
> 
> In that code\, steps 2 and 3 happen first\, and then all three steps
> happen at the same time inside the eval\.  \(That the internal code that
> handles sub definitions should have to handle cloning as well really
> complicates things\.\)  This is an exception to the general rule\, which
> is why I question whether it was a good idea\.  Usually steps 2 and 3
> happen only when the enclosing block is entered\.
> 
> When you have a BEGIN block with an eval inside it\, then things get a
> little screwy\.
> 
> my sub f;
> BEGIN \{ eval 'sub f \{\.\.\.\}'; \}
> 
> Because of the way lexical\-sub\-definition\-in\-eval has been special\-
> cased to work\, you get all three steps happening at once at compile
> time\, and then steps 2 and 3 repeat at run time\.
> 
> Now this may all sound like internal jargon\, but it is all observable
> behaviour\.  Steps 2 and 3 always happen together with my\-subs\, but
> step 1 happens separately\.  One can observe step 1 through attributes
> \(i\.e\.\, based on when MODIFY\_CODE\_ATTRIBUTES is called\.\)  One can
> observe steps 2 and 3 \(actually a single step\) by the effects of
> calling the sub\.

Oh\, I think I know what I was missing\.  When the sub is defined in a string eval and declared outside the eval \(this is the special case referred to above\)\, the sub is immediately cloned and the clone is installed in the pad\.  Otherwise\, the protosub is set aside for the sake of future cloning\.  Maybe we need \*both\* to happen in string eval\, and everybodyâs expectations will be met\.  Or maybe not everybodyâs\.

&nbsp;    for \(1\,2\) \{
&nbsp;        my sub f \{ print "42\\n" \};
&nbsp;        f\(\);
&nbsp;        eval 'sub f \{ print "43\\n" \}';
&nbsp;        f\(\);
&nbsp;    \}

That would print&#8203;:

42
43
43
43

But it would allow BEGIN\-time string eval to set things up properly\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jan 07 22&#8203;:08&#8203;:03 2015\, aristotle wrote&#8203;:
> \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-04 19&#8203;:45\]&#8203;:
> > $ \./perl \-Ilib \-e 'use feature "&#8203;:all"; my sub f; BEGIN \{ eval "sub
> > f\{warn 42\}"\} f'
> > The lexical\_subs feature is experimental at \-e line 1\.
> > Undefined subroutine &f called at \-e line 1\.
> 
> â¦ this \*ought\* to work\.

And it started working with commit v5\.21\.6\-197\-g0f94cb1\, but I do not know why\.  \(That commit was not supposed to change behaviour at all\.\)  It needs a test\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed May 18 22&#8203;:22&#8203;:33 2016\, sprout wrote&#8203;:
> On Wed Jan 07 22&#8203;:08&#8203;:03 2015\, aristotle wrote&#8203;:
> > \* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2014\-12\-04 19&#8203;:45\]&#8203;:
> > > $ \./perl \-Ilib \-e 'use feature "&#8203;:all"; my sub f; BEGIN \{ eval "sub
> > > f\{warn 42\}"\} f'
> > > The lexical\_subs feature is experimental at \-e line 1\.
> > > Undefined subroutine &f called at \-e line 1\.
> >
> > â¦ this \*ought\* to work\.
> 
> And it started working with commit v5\.21\.6\-197\-g0f94cb1\, but I do not
> know why\.  \(That commit was not supposed to change behaviour at all\.\)
> It needs a test\.

Much of what I said in this ticket was based on a misunderstanding of how newMYSUB determines how to install the sub\.Â¹  Whether a newly\-defined lexical sub is set aside for cloning later or cloned and installed in the lexical scope immediately is determined by whether the scope in which it is declaredÂ² is currently active \(the sub at that scope level is running\)\.

So\, that

&nbsp;    my sub f;
&nbsp;    BEGIN \{ eval "sub f \{\.\.\.\}" \}

did not work was not due to the model followed\, but simply due to a bug in the code\.  I do not think it is worth poring through the code to find out exactly why it used to fail\.  But it probably had something to do with the more complex structure of SV \+ magic not being handled correctly somewhere\.  v5\.21\.6\-197\-g0f94cb1 simplified the data structure and apparently fixed the mishandling thereof at the same time\.

All this is to say that this bug can be closed and that the current model\, which I suggested changing\, is fine\.  I added a test in 38fe8a04ecd\.


Â¹I know\, I even \*wrote\* the code\.  But by the time this ticket came up\, I had forgotten it\.  Also\, just \*look\* at the code in newMYSUB\.  Thatâs enough to hurt anyoneâs head\.  Maybe I should plead insanity\, but not my own in sanity; rather\, insanity in Perlâs run\-time\-vs\-compile\-time scoping model\.

Â²Note the distinction between declaration and definition\.  The former refers to âmy subâ\, which establishes the lexical scope\.  The latter refers to the body\, which defines what the sub does\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
